src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,320 @@
+        onChange: handleChange,
+        onUpdate,
+        min: externalMin,
+        max: externalMax,
+        shiftMultiplier: externalShiftMultiplier = 10,
+        step: externalStep = 1,
+        size,
+        view,
+        disabled,
+        hasControls = true,
+        validationState,
+        onFocus,
+        onBlur,
+        onKeyDown,
+        onKeyUp,
+        allowMouseWheel = false,
+        allowDecimal = false,
+        className,
+    } = {...defaultBaseTextInputProps, ...props};
+
+    const {
+        min,
+        max,
+        step: baseStep,
+        value,
+        defaultValue,
+        shiftMultiplier,
+    } = getInternalState({
+        min: externalMin,
+        max: externalMax,
+        step: externalStep,
+        shiftMultiplier: externalShiftMultiplier,
+        allowDecimal,
+        value: externalValue,
+        defaultValue: externalDefaultValue,
+    });
+
+    const [inputValue, setInputValue] = React.useState(
+        getStringValue(value) ?? getStringValue(defaultValue),
+    );
+
+    React.useEffect(() => {
+        const stringPropsValue = getStringValue(value);
+        setInputValue((currentInputValue) => {
+            if (!areStringRepresentationOfNumbersEqual(currentInputValue, stringPropsValue)) {
+                return stringPropsValue;
+            }
+            return currentInputValue;
+        });
+    }, [value]);
+
+    const clamp = true;
+    const [step, setStep] = React.useState(baseStep);
+    const [active, setActive] = React.useState(false);
+    const safeValue = value ?? 0;
+
+    const state = getInputControlState(validationState);
+
+    const canIncrementNumber =
+        clampToNearestStepValue({value: safeValue + step, step, min, max}) > safeValue;
+
+    const canDecrementNumber =
+        clampToNearestStepValue({value: safeValue - step, step, min, max}) < safeValue;
+
+    const innerControlRef = React.useRef<HTMLInputElement>(null);
+    const fieldRef = useFormResetHandler({
+        initialValue: defaultValue,
+        onReset: onUpdate ?? voidFunction,
+    });
+    const handleRef = useForkRef(props.controlRef, innerControlRef, fieldRef);
+
+    const handleIncrement = () => {
+        if (canIncrementNumber) {
+            onUpdate?.(
+                clampToNearestStepValue({value: safeValue + step, step: baseStep, min, max}),
+            );
+        }
+    };
+
+    const handleDecrement = () => {
+        if (canDecrementNumber) {
+            onUpdate?.(
+                clampToNearestStepValue({value: safeValue - step, step: baseStep, min, max}),
+            );
+        }
+    };
+
+    const handleWheel: React.WheelEventHandler<HTMLInputElement> = (e) => {
+        const delta = e.shiftKey ? e.deltaX : e.deltaY;
+        if (delta > 0) {
+            handleIncrement();
+        } else if (delta < 0) {
+            handleDecrement();
+        }
+    };
+
+    const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
```

Comment:
If min/max are set, we can set the value to min/max on Home/End.

Reply:
But in native input type=number on End/Home just the caret moves to the end/beginning of the content...
Wouldn't this difference in behaviour with native control be confusing?